### PR TITLE
boards/nucleo-l496zg: add initial support

### PR DESCRIPTION
**Pull request by aabadie**
Port of ST [Nucleo-L496ZG](http://www.st.com/en/evaluation-tools/nucleo-l496zg.html) just made by following the datasheets.
If someone with the board could test that would be great.

### CODE REVIEW

File: boards/nucleo144-common/include/board_common.h [hunk 1]
```diff
@@ -35,11 +35,18 @@ extern "C" {
...
+#if defined(STM32L496ZG)
+#define LED0_PORT           PORT_C
+#define LED0_PIN_NUM        7
+#else
+#define LED0_PORT           PORT_B
+#define LED0_PIN_NUM        0
+#endif
+#define LED0_PIN            GPIO_PIN(LED0_PORT, LED0_PIN_NUM)
+#define LED0_MASK           (1 << LED0_PIN_NUM)
+#define LED0_ON             (LED0_PORT->BSRR = LED0_MASK)
```
Comment (haukepetersen) [inline boards/nucleo144-common/include/board_common.h]: this seems broken now: `LED0_PORT` is either `PORT_C` or `PORT_B`, where it should be translated to `GPIOC` and `GPIOB`, respectively.

File: boards/nucleo144-l496/include/periph_conf.h [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+ * @{
+ */
+#define ADC_NUMOF           (0)
+/** @} */
+
+/**
+ * @name    DAC configuration
+ * @{
+ */
+#define DAC_NUMOF           (0)
```
Comment (haukepetersen) [inline boards/nucleo144-l496/include/periph_conf.h]: remove this block, not needed anymore.

File: cpu/stm32l4/include/cpu_conf.h [hunk 3]
```diff
@@ -42,8 +44,11 @@ extern "C" {
 #define CPU_DEFAULT_IRQ_PRIO            (1U)
 #if defined(STM32L432KC)
 #define CPU_IRQ_NUMOF                   (82U)
+#elif defined(STM32L496ZG)
+#define CPU_IRQ_NUMOF                   (91U)
```
Comment (haukepetersen) [inline cpu/stm32l4/include/cpu_conf.h]: you add only 8 additional irq vectors to the vector table, but increment the number of available IRQs by 10 - I have the impression something got out of sync here :-)

Comment (aabadie) [inline cpu/stm32l4/include/cpu_conf.h]: you are right, comparing with the others, I could find the problem: IRQ_NUMOF incorrect (should be 90) + one missing in vectors.c.

**Comment by aabadie** [comment]
@haukepetersen, comments addressed.

File: boards/nucleo144-common/include/arduino_pinmap.h [hunk 4]
```diff
@@ -59,7 +59,7 @@ extern "C" {
 #define ARDUINO_PIN_A0          GPIO_PIN(PORT_A, 3)
 #define ARDUINO_PIN_A1          GPIO_PIN(PORT_C, 0)
 #define ARDUINO_PIN_A2          GPIO_PIN(PORT_C, 3)
-#if defined(CPU_MODEL_STM32F413ZH) || defined(CPU_MODEL_STM32F412ZG)
+#if defined(CPU_MODEL_STM32F413ZH) || defined(CPU_MODEL_STM32F412ZG) || defined(CPU_MODEL_STM32L496ZG)
```
Comment (haukepetersen) [inline boards/nucleo144-common/include/arduino_pinmap.h]: line length...

File: boards/nucleo144-common/include/board_common.h [hunk 5]
```diff
@@ -36,11 +36,18 @@ extern "C" {
...
-#define LED0_OFF            (GPIOB->BSRR = (LED0_MASK << 16))
-#define LED0_TOGGLE         (GPIOB->ODR  ^= LED0_MASK)
+#if defined(STM32L496ZG)
+#define LED0_PORT           GPIOC
+#define LED0_PIN_NUM        7
+#else
+#define LED0_PORT           GPIOB
+#define LED0_PIN_NUM        0
+#endif
+#define LED0_PIN            GPIO_PIN(LED0_PORT, LED0_PIN_NUM)
```
Comment (haukepetersen) [inline boards/nucleo144-common/include/board_common.h]: Now you broke this, the right thing here would be `GPIO_PIN(PORT_C, LED0_PIN_NUM)`...

**Comment by aabadie** [comment]
> two more minor issues

addressed

**Comment by aabadie** [comment]
rebased and squashed. Anyone with the hardware for testing ?

**Comment by haukepetersen** [comment]
overlooked that comment... Anyway, I think the chance of someone having this boards is not that great -> so either we merge blindly for now, or someone has to buy that board and we wait for that.

**Comment by aabadie** [comment]
@haukepetersen, I could test this one today and found remaining issues. I could fix some but now I'm stuck with the debug UART:
The problem is that this board (and other new nucleo L4) are using LPUART1 with ST-Link. LPUART1 is on another clock configuration register and there might be other different things with uart driver implementation.
Do you have any experience with such peripheral ?

**Comment by haukepetersen** [comment]
> Do you have any experience with such peripheral ?

partly. I think we need to multiplex the uart interface internally for both types of UART devices for now.

**Comment by aabadie** [comment]
@haukepetersen, to move forward here (until there's support lpuart on stm32), I mapped the stdio UART to the pin on the arduino connector (similar to what I did in #7491).

Everything is working now (didn't test SPI though): timer, uart, rtt, rtc, pwm.

If your ACK still holds and you agree with the stdio uart mapping, this one is good to go I think.

**Comment by aabadie** [comment]
@astralien3000 has the hardware so maybe he can confirm that it works with the usual on-board tests.

**Comment by aabadie** [comment]
rebased, @astralien3000, interested in giving this one a try ?

**Comment by aabadie** [comment]
> I can't flash. I have no time to fix that for the moment.

no problem, I somehow failed with the interrupt vectors when rebasing the branch. Will check later.

**Comment by aabadie** [comment]
@astralien3000, I tested this branch again and it works like a charm. You may need to update your openocd version. I'm using the very latest one (master branch):
```
$ openocd --version
Open On-Chip Debugger 0.10.0+dev-00399-g09076d10 (2018-04-12-15:43)
```

**Comment by astralien3000** [comment]
I also needed to update gcc to gcc-7.
uart, xtimer, pwm, leds and button are working, but the test periph_timer hardfault for TIMER_1.

**Comment by aabadie** [comment]
>  the test periph_timer hardfault for TIMER_1.

@astralien3000, I found remaining problems in the vectors.c file. They should be fixed now.

**Comment by astralien3000** [comment]
Now everything is working, you can squash.

**Pull request merged**